src/Open3D/Visualization/Rendering/Filament/FilamentRenderToBuffer.cpp

```diff
@@ -0,0 +1,179 @@
+    renderer_ = engine_.createRenderer();
+}
+
+FilamentRenderToBuffer::~FilamentRenderToBuffer() {
+    engine_.destroy(swapchain_);
+    engine_.destroy(renderer_);
+
+    if (buffer_) {
+        free(buffer_);
+        buffer_ = nullptr;
+
+        bufferSize_ = 0;
+    }
+
+    if (parent_) {
+        parent_->OnBufferRenderDestroyed(this);
+        parent_ = nullptr;
+    }
+}
+
+void FilamentRenderToBuffer::SetDimensions(const std::size_t width,
+                                           const std::size_t height) {
+    if (swapchain_) {
+        engine_.destroy(swapchain_);
+    }
+
+    swapchain_ = engine_.createSwapChain(width, height,
+                                         filament::SwapChain::CONFIG_READABLE);
+    view_->SetViewport(0, 0, width, height);
+
+    width_ = width;
+    height_ = height;
+
+    bufferSize_ = width * height * 3 * sizeof(std::uint8_t);
+    if (buffer_) {
+        buffer_ = static_cast<std::uint8_t*>(realloc(buffer_, bufferSize_));
+    }
+}
+
+void FilamentRenderToBuffer::CopySettings(const View* view) {
+    auto* downcast = dynamic_cast<const FilamentView*>(view);
+    // NOTE: This class used to copy parameters from the view into a view
+    // managed by this class. However, the copied view caused anomalies when
+    // rendering an image for export. As a workaround, we keep a pointer to the
+    // original view here instead.
+    view_ = const_cast<FilamentView*>(downcast);
+    if (downcast) {
+        auto vp = view_->GetNativeView()->getViewport();
+        SetDimensions(vp.width, vp.height);
+    }
+}
+
+View& FilamentRenderToBuffer::GetView() { return *view_; }
+
+using PBDParams = std::tuple<FilamentRenderToBuffer*,
+                             FilamentRenderToBuffer::BufferReadyCallback>;
+void FilamentRenderToBuffer::RequestFrame(Scene* scene,
+                                          BufferReadyCallback callback) {
+    if (!scene) {
+        utility::LogDebug(
+                "No Scene object was provided for rendering into buffer");
+        callback({0, 0, nullptr, 0});
+        return;
+    }
+
+    if (pending_) {
+        utility::LogWarning(
+                "Render to buffer can process only one request at time");
+        callback({0, 0, nullptr, 0});
+        return;
+    }
+
+    pending_ = true;
+
+    if (buffer_ == nullptr) {
+        buffer_ = static_cast<std::uint8_t*>(malloc(bufferSize_));
+    }
+
+    callback_ = callback;
+}
+
+void FilamentRenderToBuffer::ReadPixelsCallback(void*, size_t, void* user) {
+    auto params = static_cast<PBDParams*>(user);
+    FilamentRenderToBuffer* self;
+    BufferReadyCallback callback;
+    std::tie(self, callback) = *params;
+
+    callback({self->width_, self->height_, self->buffer_, self->bufferSize_});
+
+    self->frameDone_ = true;
+}
+
+void FilamentRenderToBuffer::Render() {
+    bool shotDone = false;
+    frameDone_ = false;
+    while (!frameDone_) {
+        if (renderer_->beginFrame(swapchain_)) {
```

Comment:
Filament's beginFrame() returns false if it wants to skip a frame. Rendering is done on a separate thread, so render calls return immediately but nothing actually happens until later, at which point the callback is called. I suspect that the loop is only to ensure that *if* beginFrame() returns false, we continue to try until it returns true, but that in actuality, it always returns true the first time. I have no idea what thread gets called back, and the Filament documentation is ... limited, but I don't see how it could be anything but the render thread, and the error printed on macOS seems to confirm this.